### ISSUE TITLE
nondeterministic segfault

### ISSUE THREAD

**Issue by lukemetz** [dmlc/mxnet#1242]
Finally was able to get a stacktrace out of this nondeterministic crash that occurs when MXNet is used. (normally running with gdb fixes the problem...) I am not sure what causes it, but rebooting appears to make it happen less (which is odd). It seems to show up when making a large number of executors and never when training a big model.

This occurred at: ee4933864263d5e815b9aa56c00b89a81f568a1e

Stacktrace from gdb bellow:
```
#0  __GI___libc_free (mem=0x8d4804a8147502a8) at malloc.c:2929
#1  0x00007fffcd0bbeb3 in std::vector<mxnet::NDArray, std::allocator<mxnet::NDArray> >::~vector() ()
   from /home/luke/Repos/luke_mxnet/python/mxnet/../../lib/libmxnet.so
#2  0x00007fffcd25880d in std::_Function_base::_Base_manager<mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}>::_M_manager(std::_Any_data&, std::_Function_base::_Base_manager<mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}> const&, std::_Manager_operation) ()
   from /home/luke/Repos/luke_mxnet/python/mxnet/../../lib/libmxnet.so
#3  0x00007fffcd0dbcf1 in std::_Function_base::_Base_manager<void mxnet::Engine::PushSync<mxnet::engine::ThreadedEngine::DeleteVariable(std::function<void (mxnet::RunContext)>, mxnet::Context, mxnet::engine::Var*)::{lambda(mxnet::RunContext)#1}>(mxnet::engine::ThreadedEngine::DeleteVariable(std::function<void (mxnet::RunContext)>, mxnet::Context, mxnet::engine::Var*)::{lambda(mxnet::RunContext)#1}, mxnet::Context, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var*> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var*> > const, mxnet::FnProperty, int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}>::_M_manager(std::_Any_data&, void mxnet::Engine::PushSync<mxnet::engine::ThreadedEngine::DeleteVariable(std::function<void (mxnet::RunContext)>, mxnet::Context, mxnet::engine::Var*)::{lambda(mxnet::RunContext)#1}>(mxnet::engine::ThreadedEngine::DeleteVariable(std::function<void (mxnet::RunContext)>, mxnet::Context, mxnet::engine::Var*)::{lambda(mxnet::RunContext)#1}, mxnet::Context, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var*> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var*> > const, mxnet::FnProperty, int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1} const&, std::_Manager_operation) ()
   from /home/luke/Repos/luke_mxnet/python/mxnet/../../lib/libmxnet.so
#4  0x00007fffcd0e1e7f in mxnet::engine::ThreadedEngine::OnComplete(mxnet::engine::ThreadedOpr*) ()
   from /home/luke/Repos/luke_mxnet/python/mxnet/../../lib/libmxnet.so
#5  0x00007fffcd2573ea in std::_Function_handler<void (mxnet::RunContext, mxnet::engine::CallbackOnComplete), mxnet::GraphExecutor::GetOpExecEntry(unsigned int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}>::_M_invoke(std::_Any_data const&, mxnet::RunContext, mxnet::engine::CallbackOnComplete) ()
   from /home/luke/Repos/luke_mxnet/python/mxnet/../../lib/libmxnet.so
#6  0x00007fffcd0ec339 in mxnet::engine::ThreadedEngine::ExecuteOprBlock(mxnet::RunContext, mxnet::engine::OprBlock*) ()
   from /home/luke/Repos/luke_mxnet/python/mxnet/../../lib/libmxnet.so
#7  0x00007fffcd0ec740 in std::_Function_handler<void (), mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, ---Type <return> to continue, or q <return> to quit---
bool)::{lambda()#3}::operator()() const::{lambda()#1}>::_M_invoke(std::_Any_data const&) ()
   from /home/luke/Repos/luke_mxnet/python/mxnet/../../lib/libmxnet.so
#8  0x00007fffc20e0a40 in ?? ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#9  0x00007ffff7bc4182 in start_thread (arg=0x7fff8da20700)
    at pthread_create.c:312
#10 0x00007ffff78f147d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

**Comment by piiswrong** [comment]
I suspect it's some threading caused memory corruption.
Could you verify if it happens when using NaiveEngine?
By making a large number of executors, do you mean using them in parallel or creating new ones after destroying old ones?

**Comment by lukemetz** [comment]
Now I can't seem to reproduce anything. If It happens again I will swap engines and see.
I am not using them in parallel on purpose. I would not be surprised if MXNet does have some threaded activity happening. I am creating a new executor with bind, immediately running forward, then grabbing the output variable and calling asnumpy(). The executor then goes out of scope for python to theoretically deal with it.
I am not sure if these issues are meant to be purely reproducible or not. If not this can be closed.

**Comment by lukemetz** [comment]
I was able to today get it to be semi reproducible. Moving to the Naive Engine did seem to fix the problem at least as far as I can tell.

**Comment by tqchen** [comment]
closing for now, feel free to reopen when we get back to this issue
